src/Symfony/Contracts/Translation/ParameterInterface.php

```diff
@@ -0,0 +1,24 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Contracts\Translation;
+
+/**
+ * Implementation of the ICU recommendation to first format advanced parameters before translation.
+ *
+ * @see https://unicode-org.github.io/icu/userguide/format_parse/messages/#format-the-parameters-separately-recommended
+ *
+ * @author Sylvain Fabre <syl.fabre@gmail.com>
+ */
+interface ParameterInterface
+{
+    public function format(string $locale = null): string;
```

Comment:
This means the parameter should handle (in a singleton way or re-creating an instance).

Here for instance, we create a `new NumberFormatter()` to format number, and it decides to use `.` or `,` for decimal separator. This class cannot be changed/extended and it left no room for customization or adding languages.

I think a parameter you translate is actually a translatable thing so you should pass to it a `$translator` and translate it like any other `TranslatableInterface`:
```php
public function trans(TranslatorInterface $translator, string $locale = null): string;
```

This way, the parameter can (if it needs to) get the decimal separator, or formatting template from the translations provided by the `$translator`, so you can control it from your translations files (pot, xliff, etc.)

Reply:
@kylekatarnls As explained at https://github.com/symfony/symfony/pull/41858#issuecomment-870417785 I don't think that a parameter should be translated that's why I'm proposing another interface without the `TranslatorInterface` argument.

Reply:
First of all, there is no real overlap between our 2 PRs, when I have dates to translate I use `Carbon` so it uses the `TranslatorInterface` to pick the per-language config. If I want to pass a number/currency/date to a `TranslatableMessage` I will simply wrap it into an other custom `TranslatableInterface` so I can mock the behavior of `TranslatorInterface` in my test and I control the whole localization from that service.

If we receive only the `$locale` we don't have any config/context, only a pure object that needs its own logic to decide what to do for each locale. That's a different approach, that's not the one I will use in my apps, but I see that can be quite good in some situation, it's perfectly fine for me to propose both to users:
- be able to pass a `ParameterInterface` as a parameter that get automatically formatted according to the locale
- be able to pass a `TranslatableInterface` as a parameter that get automatically formatted according to the locale and the translator that was used (and so  eventually the default translator locale if $locale is null)